File: LeetCode/1882. Process Tasks Using Servers/solution.py
class Solution:
    def assignTasks(self, servers: List[int], tasks: List[int]) -> List[int]:
        n, m = len(servers), len(tasks)
        q = []
        ready = [(w,i) for i, w in enumerate(servers)]
        heapq.heapify(ready)
        ans = [-1]*m 
        start = 0
        for i in range(m):
            start = max(start, i) 
            if not ready:
                start = q[0][0]
            while q and start >= q[0][0]:
                _, w, idx = heapq.heappop(q)
                heapq.heappush(ready, (w,idx))
            w, idx = heapq.heappop(ready)
            ans[i] = idx
            heapq.heappush(q, (start+tasks[i], w, idx))
        return ans

File: LeetCode/1882. Process Tasks Using Servers/solution.cpp
class Solution {
public:
    vector<int> assignTasks(vector<int>& servers, vector<int>& tasks) {
        int n = servers.size(), m = tasks.size();
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> available, todo;
        int cur = 0, clock = 0;
        vector<int> ans(m);
        for (int i = 0; i < n; i++){
            available.push({servers[i],i});
        }
        while (cur < m){
            while (!todo.empty() && todo.top().first == clock){
                auto [t,idx] = todo.top();todo.pop();
                available.push({servers[idx],idx});
            }
            while (cur <= min(clock, m-1) && !available.empty()){
                auto [w, idx] = available.top();available.pop();
                todo.push({clock+tasks[cur],idx});
                ans[cur++] = idx;
            }
            clock = clock < m ? clock+1 : todo.top().first;
        }
        return ans;
    }
};